Dockerfile

```diff
@@ -1,51 +1,116 @@
+### Base image ##
+# All package installation, updates, etc., anything with APT should be done here in a single step
+# hadolint ignore=DL3006
+FROM ${BASE_IMAGE}@${BASE_DIGEST} as base
+
+WORKDIR /
+
+# Upgrade all outdated packages and install missing ones (e.g. locales, tini)
+# This breaks reproducibility of builds, but is acceptable to gain access to security patches faster
+# than the base image releases
+# FYI, installing packages via APT also updates the dpkg files, which are few MBs, but removing or
+# caching them could break stuff (like not knowing the package is present) or container scanners
+# hadolint ignore=DL3008
+RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
+    --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    --mount=type=cache,target=/var/log/apt,sharing=locked \
     apt-get -qq update && \
-    apt-get install -y --no-install-recommends tini=0.19.0-1 && \
-    cp /usr/bin/tini .
-COPY --link --chown=1000:0 docker/utils/startup.sh .
+    apt-get install -yqq --no-install-recommends tini ca-certificates && \
+    apt-get upgrade -yqq --no-install-recommends
+
+
+### Build custom JRE using the base JDK image
+# hadolint ignore=DL3006
+FROM ${JDK_IMAGE}@${JDK_DIGEST} as jre-build
+
+# Build a custom JRE which will strip down and compress modules to end up with a smaller Java \
+# distribution than the official JRE. This will also include useful debugging tools like
+# jcmd, jmod, jps, etc., which take little to no space. Anecdotally, compressing modules add around
+# 10ms to the start up time, which is negligible considering our application takes ~10s to start up.
+# See https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/
+# hadolint ignore=DL3018
+RUN jlink \
+     --add-modules ALL-MODULE-PATH \
+     --strip-debug \
+     --no-man-pages \
+     --no-header-files \
+     --compress=2 \
+     --output /jre && \
+   rm -rf /jre/lib/src.zip
+
+### Java base image
+FROM base AS java
+WORKDIR /
+
+# Inherit from previous build stage
+ARG JAVA_HOME=/opt/java/openjdk
+
+# Default to UTF-8 file encoding
+ENV LANG='C.UTF-8' LC_ALL='C.UTF-8'
+
+# Setup JAVA_HOME and binaries in the path
+ENV JAVA_HOME ${JAVA_HOME}
+ENV PATH $JAVA_HOME/bin:$PATH
+
+# Copy JRE from previous build stage
+COPY --from=jre-build /jre ${JAVA_HOME}
+
+# https://github.com/docker-library/openjdk/issues/212#issuecomment-420979840
+# https://openjdk.java.net/jeps/341
+# TL;DR generate some class data sharing for faster load time
+RUN java -Xshare:dump;
 
 ### Build zeebe from scratch ###
-FROM maven:3-eclipse-temurin-17 as build
+FROM java as build
 WORKDIR /zeebe
 ENV MAVEN_OPTS -XX:MaxRAMPercentage=80
 COPY --link . ./
-RUN --mount=type=cache,target=/root/.m2,rw mvn -B -am -pl dist package -T1C -D skipChecks -D skipTests
-RUN mv dist/target/camunda-zeebe .
+RUN --mount=type=cache,target=/root/.m2,rw \
+    ./mvnw -B -am -pl dist package -T1C -D skipChecks -D skipTests && \
```

Comment:
:x: `--build-arg DIST="build"`  fails due to the absence of `mvnw`

```
 > [build 3/3] RUN --mount=type=cache,target=/root/.m2,rw     ./mvnw -B -am -pl dist package -T1C -D skipChecks -D skipTests &&     mv dist/target/camunda-zeebe .:
0.300 /bin/sh: 1: ./mvnw: not found
```

but then `mvn` is also not available, shall we consider backportint the maven wrapper as well? :thinking:

Reply:
We can use the Maven image, it doesn't really matter that much, though it was nice that we reused the same JDK to build :shrug: